### ISSUE TITLE
Fix incorrect config when using standalone Stylelint

### ISSUE THREAD

**Issue by romainmenke** [stylelint/stylelint#8014]
https://github.com/stylelint/stylelint/pull/7826 introduced a regressions that causes Stylelint to run with unexpected configs.

This was surfaced through the [ecosystem checks](https://github.com/stylelint/stylelint-ecosystem-tester/actions/runs/11116858000) on this package: https://github.com/pascalduez/stylelint-config-css-modules

**To replicate:**

- have a stylelint config file in the root of your project
- run stylelint with a config object different from the config file in the root

```js
await stylelint.lint({
    config: {
      extends: [
        'stylelint-config-standard',
        path.join(import.meta.dirname, '..', 'index.js'),
      ],
    },
    files: 'test/*.scss',
  });
```

**I expect** the `config` value passed directly to `stylelint.lint` to be the active config.

**Instead** the config file in the root is loaded and overrides any config set manually.

Minimal reproduction:
- clone https://github.com/pascalduez/stylelint-config-css-modules
- run `npm install --no-audit --no-save "stylelint/stylelint"`
- run `npm test`

-----

I don't know enough about the config loading code to figure out more about this issue.
And sorry for not following a template :) The issue didn't fit the regular pattern.

**Comment by ybiquitous** [comment]
@romainmenke Thanks for the report. I'll look into it. 👍🏼

**Comment by ybiquitous** [comment]
In `getConfigForFile()`, returning cache seems incorrect:

https://github.com/stylelint/stylelint/blob/006cf1fa077f0cbff244300c523007a794fd2a17/lib/getConfigForFile.mjs#L33-L35

**Comment by ybiquitous** [comment]
I have taken more time than I thought. Perhaps, there seems to be a problem with the caching of Stylelint config objects.

**Comment by romainmenke** [comment]
@ybiquitous have you been able to write a failing test?
I just tried and couldn't replicate in the way I expected it to.
Maybe I am overlooking something?

**Comment by romainmenke** [comment]
```js
it('foo', async () => {
	const { report } = await standalone({
		config: {
			extends: [
				path.join(__dirname, 'fixtures', 'config-block-empty-ok.mjs')
			],
		},
		// Set CWD to a directory that contains a config with `block-no-empty: true`
		cwd: path.join(__dirname, 'fixtures', 'getConfigForFile', 'a'),
		// Set files to a child of `cwd`
		files: path.join(__dirname, 'fixtures', 'getConfigForFile', 'a', 'b', 'foo.css'),
	});

	expect(report).not.toContain('Unexpected empty block');
});
```

```js
// lib/__tests__/fixtures/config-block-empty-ok.mjs
export default {
	rules: {
		'block-no-empty': null,
	},
};
```

I expected this to mirror: https://github.com/pascalduez/stylelint-config-css-modules/blob/66f837854875478f2a313b4a76b0d0744934a71e/test/scss.js#L14-L17

**Comment by romainmenke** [comment]
I agree @Mouvedia, I think that changing to three `formatter: await getFormatter(stylelint),` calls would be fine to restore the previous behavior.

That would unblock releases.

I am a bit concerned that there is a larger issue, but we can investigate further later?

**Comment by Mouvedia** [comment]
> but we can investigate further later?

It depends, @ybiquitous do you want to investigate or do you wanna release right away?
Couldn't we have at least one regression test? Or is stylelint-ecosystem-tester enough?

**Comment by ybiquitous** [comment]
Thanks for the help. I want to look into this issue more.

Since multiple calls resolve this, I still guess the current caching may have a problem. I think this is a good timing to resolve a potential bug.

**Comment by romainmenke** [comment]
Sounds good!
Thank you @ybiquitous 🙇

**Comment by ybiquitous** [comment]
I just opened PR #8030 to resolve this issue!
